src/main/scala/com/codacy/parsers/implementation/CloverParser.scala

```diff
@@ -0,0 +1,71 @@
+package com.codacy.parsers.implementation
+
+import java.io.File
+
+import com.codacy.api.{CoverageFileReport, CoverageReport}
+import com.codacy.parsers.CoverageParser
+import com.codacy.parsers.util.{TextUtils, XMLoader}
+
+import scala.util.{Failure, Success, Try}
+import scala.xml.{Elem, Node, NodeSeq}
+
+object CloverParser extends CoverageParser {
+  private val COVERAGE = "coverage"
+  private val PROJECT = "project"
+  private val METRICS = "metrics"
+
+  override val name: String = "Clover"
+
+  override def parse(rootProject: File, reportFile: File): Either[String, CoverageReport] = {
+    val report = Try(XMLoader.loadFile(reportFile)) match {
+      case Success(xml) if hasCorrectSchema(xml) =>
+        Right(xml \\ COVERAGE)
+
+      case Success(_) =>
+        Left(s"Invalid report. Could not find tag hierarchy <$COVERAGE> <$PROJECT> <$METRICS> tags.")
+
+      case Failure(ex) =>
+        Left(s"Unparseable report. ${ex.getMessage}")
+    }
+
+    report.right.flatMap(parse(rootProject, _))
+  }
+
+  private def hasCorrectSchema(xml: Elem) =
+    (xml \\ COVERAGE \ PROJECT \ METRICS).nonEmpty
+
+  private def parse(rootProject: File, report: NodeSeq): Either[String, CoverageReport] = {
+    // global metrics
+    val metrics = report \ PROJECT \ METRICS
+    val totalCoverage = getCoveragePercentage(metrics)
+
+    val rootPath = TextUtils.sanitiseFilename(rootProject.getAbsolutePath)
+    val coverageFiles = (report \\ "file").map { f =>
+      getCoverageFileReport(rootPath, f)
+    }
+
+    Right(CoverageReport(totalCoverage, coverageFiles))
+  }
```

Comment:
Why returning `Either` if you always return `Right`?
```suggestion
  private def parse(rootProject: File, report: NodeSeq): CoverageReport = {
    // global metrics
    val metrics = report \ PROJECT \ METRICS
    val totalCoverage = getCoveragePercentage(metrics)

    val rootPath = TextUtils.sanitiseFilename(rootProject.getAbsolutePath)
    val coverageFiles = (report \\ "file").map { f =>
      getCoverageFileReport(rootPath, f)
    }

    CoverageReport(totalCoverage, coverageFiles)
  }
```

Reply:
Done ✅